### DISCUSSION TITLE
Is it possible to define a List in a Schema?

### DISCUSSION THREAD

**Discussion by Florenzen68** [apple/pkl#515] [General]
This is a test .pkl file that I use in my Bachelor Thesis:
```
// Configuration for a hypothetical building automation system
amends "BuildingSchema.pkl" // Needed for validation purpose

building {
  floors { // Define the floors in the building
    new { // Adding a new floor
      floor = 1 // Specify the floor number
      rooms { // Define the rooms on this floor
        new { // Adding a new room
          name = "Conference Room"
          lights { // Define the lights in this room
            new { // Adding new lights
              id = 1
              type = "LED"
              status = "off"
              brightness = 75
            }
          }
        }
        new { // Adding another new room
          name = "Office 101"
          lights {
            new {
              id = 2
              type = "LED"
              status = "off"
              brightness = 50
            }
          }
        }
        new { // Adding another new room
          name = "Lobby"
          lights {
            new {
              id = 3
              type = "LED"
              status = "on"
              brightness = 65
            }
          }
        }
      }
    }
  }
}
```

However I want to validate the .pkl with this "Schema":
```
module BuildingSchema

building: Building

class Building {
  floors: List(Floor)
}

class Floor {
  floor: UInt16
  rooms: List(Room)
}

class Room {
  name: String(!isEmpty)
  lights: List(Light)
}

class Light {
  id: UInt16
  type: String(!isEmpty)
  status: "on" | "off"
  brightness: UInt8(isBetween(0, 100))
}
```

But obviously I get an error doing this. Is there a way to support "Lists" with a certain type? I tried to use Listing and Dynamic as Types but then the other rules defined in the classes got ignored. Can someone help?

**Comment by Florenzen68** [comment]
FYI: I made a mistake and used "(" and ")" instead of "<" and ">" while trying the Listing solution.

**Reply by bioball** [reply]
Glad to see that you answered your own question.

FYI, configuration data should use [Listing](https://pkl-lang.org/main/current/language-reference/index.html#listings), because they are regular objects that can be amended. [List](https://pkl-lang.org/main/current/language-reference/index.html#lists) is a type that is meant for in-language operations that are more "algorithm-y".

**Reply by Florenzen68** [reply]
Thank you very much for the explanation and reference. I was searching in the blog entry and not using the user manual 🤦
